tunnel/intra/doh/doh.go

```diff
@@ -190,34 +196,101 @@ func (e *queryError) Unwrap() error {
 // Given a raw DNS query (including the query ID), this function sends the
 // query.  If the query is successful, it returns the response and a nil qerr.  Otherwise,
 // it returns a nil response and a qerr with a status value indicating the cause.
-// Independent of the query's success or failure, this function also returns the IP
-// address of the server on a best-effort basis, returning the empty string if the address
-// could not be determined.
-func (t *transport) doQuery(q []byte) (response []byte, server string, qerr error) {
+// Independent of the query's success or failure, this function also returns the
+// address of the server on a best-effort basis, or nil if the address could not
+// be determined.
+func (t *transport) doQuery(q []byte) (response []byte, server *net.TCPAddr, qerr error) {
 	if len(q) < 2 {
 		qerr = &queryError{BadQuery, fmt.Errorf("Query length is %d", len(q))}
 		return
 	}
 	id0, id1 := q[0], q[1]
+	id := int(id0)*256 + int(id1)
```

Comment:
nit: consider using `id := binary.BigEndian.Uint16(q[:1])` and `binary.BigEndian.PutUint16(q[:1], 0)` to avoid working with single bytes.

Reply:
Good idea.  Done.